# EECS C106A Python Bootcamp
### EECS C106A: Introduction to Robotics, Fall 2025

# Introduction

We will be using the Python programming language for lab and homework assignments in EECS/BioE/MechE C106A/206A. Some homework assignments will entail matrix calculations where Python will come in handy, but you are welcome to use something like Matlab instead. This notebook is meant to be a mini bootcamp on Python for students who have had experience programming in another language already (e.g. Matlab) or need a quick refresher on Python. We will be using a few popular libraries (numpy, scipy) that are very useful. If you have experience with Matlab but not Python, we recommend checking out the [numpy for Matlab users guide](https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html).

# Table of Contents
* [IPython Basics](#IPython-Basics)
* [Python Data Types](#Data-Types)
* [Python Functions](#Functions)

# IPython Basics

For those who have never used IPython, it is a command shell for interactive Python programming. You can imagine it as allowing you to run blocks of Python code that you would execute in a single python script (python [script_name].py) in the terminal. Benefits of using IPython include easier visualization and debugging. The purpose of this bootcamp in IPython is to give you an idea of basic Python syntax and semantics. For labs you are still expected to write and execute actual Python scripts to work with ROS.

### Executing Cells

ipython notebooks are constituted of cells, that each have text, code, or html scripts. To run a cell, click on the cell and press Shift+Enter or the run button on the toolbar. Run the cell below, you should expect an output of 6:

In [3]:
1+2+3

6

### Stopping or Restarting Kernel

To debug, or terminate a process, you can interupt the kernel by clicking Kernel/interrupt on the toolbar. If interuppting doesn't work, or you would like to restart all the processes in the notebook, click Kernel/restart. Try interrupting the following block:

In [4]:
import time

while True:
    print("bug")
    time.sleep(1.5)
    break
    

bug


### Import a library

To import a certain library `x`, just type `import x`
Calling function `y` from that library is simply `x.y`
To give the library a different name (e.g. abbreviation), type `import x as z`

In [5]:
!pip3 install numpy
import numpy as np
np.add(3, 4)

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.doubanio.com/simple



[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


7

# Python

## Data Types

### Integers and Floats

In Python2, integer division returns the floor. In Python3, there is no floor unless you specify using double slahes. The differences between Python2 and Python3 you can [check out](https://wiki.python.org/moin/Python2orPython3).

In [6]:
5 / 4

1.25

In [7]:
5.0 / 4

1.25

### Booleans

Python implements all usual operators for Boolean logic. English, though, is used rather than symbols (no &, ||, etc.). Pay attention to the following syntax, try to guess what the output for each print statement should be before running the cell.

In [31]:
print(0 == False)

t = True
print(1 == t)

print(0 != t)

print(t is not 1) 
print(t == 1) 
print(t is 1) 

if t is True:
    print(0 != 0)

True
True
True
True
True
False
False


<>:8: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<>:10: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:8: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<>:10: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_37340\1957782122.py:8: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  print(t is not 1)
C:\Users\Lenovo\AppData\Local\Temp\ipykernel_37340\1957782122.py:10: SyntaxWarning: "is" with a literal. Did you mean "=="?
  print(t is 1)


### Strings

Strings are supported very well. To concatenate strings we can do the following:

In [33]:
hello = 'hello'
robot = 'robot'

print(hello + ' ' + robot + str(1))
print(f"{hello}{robot} {1}")

hello robot1
hellorobot 1


To find the length of a string use `len(...)`

In [10]:
print(len(hello + robot))

10


### Lists

A list is a mutable array of data, meaning we can alter it after insantiating it. To create a list, use the square brackets [] and fill it with elements.

Key operations:
- `'+'` appends lists
- `len(y)` to get length of list y
- `y[0]` to index into 1st element of y **Python indices start from 0
- `y[1:6]` to slice elements 1 through 5 of y

In [34]:
y = ["Robots are c"] + [0, 0, 1]
y
y1 = "Robots are c" + "".join([str(x) for x in [0, 0, 1]])
print(y1)

Robots are c001


In [12]:
len(y)

4

In [13]:
y[0]

'Robots are c'

In [14]:
# TODO: slice the first three elements of list 'y' and 
# store in a new list, then print the 2nd element of this 
# new list

### Loops

You can loop over the elements of a list like this:

In [15]:
robots = ['baxter', 'sawyer', 'turtlebot']
for robot in robots:
    print(robot)
# Prints "baxter", "sawyer", "turtlebot", each on its own line.

baxter
sawyer
turtlebot


If you want access to the index of each element within the body of a loop, use the built-in [`enumerate`](https://docs.python.org/2.7/library/functions.html#enumerate) function:

In [35]:
robots = ['baxter', 'sawyer', 'turtlebot']

# TODO: Using a for loop and the python built-in enumerate function,
# Print "#1: baxter", "#2: sawyer", "#3: turtlebot", 
# each on its own line

### Numpy Array

The numpy array is like a list with multidimensional support and more functions (which can all be found [here](https://docs.scipy.org/doc/numpy/reference/index.html)).

NumPy arrays can be manipulated with all sorts of arithmetic operations. You can think of them as more powerful lists. Many of the functions that already work with lists extend to numpy arrays.

To use functions in NumPy, we have to import NumPy to our workspace. by declaring `import numpy`, which we have done previously above in this notebook already. We typically rename `numpy` as `np` for ease of use.

### Making a Numpy Array

In [17]:
x = np.array([[1, 2, 3], [4 , 5, 6], [7, 8, 9]])
print(x)
# x is a 3x3 matrix here

[[1 2 3]
 [4 5 6]
 [7 8 9]]


### Finding the shape of a Numpy Array

In [18]:
x.shape # returns the dimensions of the numpy array

(3, 3)

### Elementwise operations

Arithmetic operations on numpy arrays correspond to elementwise operations.

In [37]:
print(x)
print
print(x * 5) # numpy carries operation on all elements!
print(x + 5)
print(x * x)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[ 5 10 15]
 [20 25 30]
 [35 40 45]]
[[ 6  7  8]
 [ 9 10 11]
 [12 13 14]]
[[ 1  4  9]
 [16 25 36]
 [49 64 81]]


### Matrix multiplication

In [20]:
print(np.dot(x, x))

[[ 30  36  42]
 [ 66  81  96]
 [102 126 150]]


### Slicing numpy arrays

Numpy uses pass-by-reference semantics so it creates views into the existing array, without implicit copying. This is particularly helpful with very large arrays because copying can be slow. Although be wary that you may be mutating an array when you don't intend to, so make sure to make a copy in these situations.

In [21]:
orig = np.array([0, 1, 2, 3, 4, 5])
print(orig)

[0 1 2 3 4 5]


Slicing an array is just like slicing a list

In [22]:
sliced = orig[1:4]
print(sliced)

[1 2 3]


Note, since slicing does not copy the array, mutating `sliced` mutates `orig`. Notice how the 4th element in `orig` changes to 9 as well.

In [ ]:
sliced[2] = 9
print(orig)
print(sliced) #view will change orig it's not a copy

[0 1 2 9 4 5]
[1 2 9]


We should use `np.copy()` to actually copy `orig` if we don't want to mutate it. 

In [38]:
orig = np.array([0, 1, 2, 3, 4, 5])
org=[0, 1, 2, 3, 4, 5]
copy = np.copy(orig)
sliced_copy = copy[1:4]
sliced_co = org[1:4]
sliced_copy[2] = 9
sliced_co[2]=9
print(orig)
print(sliced_copy)
print(org)
print(sliced_co)

[0 1 2 3 4 5]
[1 2 9]
[0, 1, 2, 3, 4, 5]
[1, 2, 9]


In [25]:
A = np.array([[5, 6, 8], [2, 4, 5], [3, 1, 10]])
B = np.array([[3, 5, 0], [3, 1, 1]])
# TODO: multiply matrix A with matrix B padded with 1's to the 
# same dimensions as A; sum this result with the identiy matrix 
# (you may find np.concatenate, np.vstack, np.hstack, or np.eye useful). 
# Make sure you don't alter the original contents of B. Print the result

### Handy Numpy function: arange

We use `arange` to instantiate integer sequences in numpy arrays. It's similar to the built-in range function in Python for lists. However, it returns the result as a numpy array, rather a simple list.

`arange(0,N)` instantiates an array listing every integer from 0 to N-1.

`arange(0,N,i)` instantiates an array listing every `i` th integer from 0 to N-1 .

In [26]:
print(np.arange(-3,4)) # every integer from -3 ... 3

[-3 -2 -1  0  1  2  3]


In [27]:
# TODO: print every other integer from 0 ... 6 multiplied by 2
# as a list

## Functions

Python functions are defined using the `def` keyword. For example:

In [41]:
def hello_robot(robot_name, yell=True):
    if yell:
        print('HELLO, %s!' % robot_name.upper())
    else:
        print('hello, %s how are you' % robot_name)

In [42]:
hello_robot('Baxter') # Prints "HELLO, BAXTER!"
hello_robot('Sawyer', yell=False)  # Prints "hello, Sawyer"

HELLO, BAXTER!
hello, Sawyer how are you


# References
- [1] EE 120 lab1
- [2] EECS 126 Lab01
- [3] EE 16a Python Bootcamp
- [4] CS 231n Python Numpy Tutorial. [Link](http://cs231n.github.io/python-numpy-tutorial/)